pg_search/src/aggregate/mod.rs

```
@@ -311,143 +469,415 @@ impl ParallelWorker for ParallelAggregationWorker<'_> {
-        let reader = SearchIndexReader::open_with_context(
-            index,
-            query.clone(),
-            false,
-            MvccSatisfies::Snapshot,
-            NonNull::new(standalone_context),
-            None,
-        )?;
-        let agg_req = serde_json::from_value(agg)?;
-        let ambulkdelete_epoch = MetaPage::open(index).ambulkdelete_epoch();
-        let segment_ids = reader
-            .segment_readers()
-            .iter()
-            .map(|r| (r.segment_id(), r.num_deleted_docs()))
-            .collect::<Vec<_>>();
+    let (reader, standalone_context, ambulkdelete_epoch, segment_ids) =
+        open_index_for_aggregation(index, qparams.base_query, MvccSatisfies::Snapshot)?;
+
+    let schema = index
+        .schema()
+        .map_err(|e| -> Box<dyn Error> { Box::new(e) })?;
+
+    let qctx = QueryContext::new(&schema, &reader, index, standalone_context);
+
+    let aggregations = build_aggregation_query(&qctx, qparams)?;
+
+    // Determine if we can use parallel execution
+    let (can_use_parallel, nworkers) = can_parallelize(&segment_ids);
+
+    // Execute aggregation (parallel or sequential)
+    if can_use_parallel {
+        // Parallel execution
         let process = ParallelAggregation::new(
             index.oid(),
-            &query,
-            &agg_req,
+            qparams,
             solve_mvcc,
             memory_limit,
             bucket_limit,
-            segment_ids,
+            segment_ids.clone(),
             ambulkdelete_epoch,
         )?;
 
-        // limit number of workers to the number of segments
-        let mut nworkers =
-            (pg_sys::max_parallel_workers_per_gather as usize).min(reader.segment_readers().len());
+        if let Some(agg_results) = execute_parallel_helper(process, nworkers)? {
+            return merge_parallel_results(aggregations, agg_results, memory_limit, bucket_limit);
+        }
+        // If parallel execution failed, fall through to sequential
```

Is this a new failure case? When will an aggregation fail when running in parallel, when it might succeed sequentially?

If this is truly a failure, then it would be better for `execute_parallel_helper` to return an error. But I also don't think that we should silently fall back to single-threaded.